text/0000-ts-adoption-plan.md

```diff
@@ -0,0 +1,780 @@
+- test infrastructure for Ember’s type, both to prevent regressions and to catch breaking changes from TypeScript early so they can be mitigated
+- release infrastructure to allow us to handle pre-release testing, feature flags, and alpha and beta releases
+- updates to our documentation to include TypeScript as a first-class citizen of the ecosystem
+
+
+### Semantic Versioning
+
+Ember packages which publish types will adhere to the [Semantic Versioning for TypeScript Types][rfc-0730] proposal. See that RFC for details on strategies for mitigating breaking changes from TypeScript, and see [**How We Teach This: Documenting SemVer**](#documenting-semver) below for discussion of how we will document this at the framework level.
+
+The 10,000-foot summary of our treatment of SemVer for TypeScript is:
+
+> Code which type checks at any given point in a major release of Ember packages will continue to type check throughout that Ember major release for consumers who use supported TypeScript versions.
+
+This is **the “no new red squiggles rule”:** if you use supported TypeScript versions, you will not see new “red squiggles” for type errors in your editor in an Ember major version.
+
+The only exceptions to the “no new red squiggles” rule are:
+
+- *Bug fixes to the types*: where we previously allowed code we should not have and which could produce runtime errors, the introduction of those red squiggles is making your code safer.
+- *Rare cases involving inferred types*: very rarely, we may make a change to the types which is safe for your code at runtime, and which allows TypeScript to detect new unused code paths or similar. These will introduce “red squiggles” in your editor, but in a way which will simply allow you to clean up code you didn’t need.
+
+For details on both of these, see [Semantic Versioning for TypeScript Types RFC][rfc-0730], which explains these in detail and provides justification for especially the second case.
+
+[rfc-0730]: https://github.com/emberjs/rfcs/pull/730
+
+
+#### Strictness
+
+In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], Ember will set `strict: true`, `noUncheckedIndexedAccess: true`. Additionally, we will set `esModuleInterop: false` and `allowSyntheticDefaultExports: false`. These will also be the defaults for Ember projects generated using a `--typescript` flag. (See discussion below under [**CLI Integration**](#cli-integration).)
+
+Beyond this, we *may* define a set of *additional* “lint”-type strictness flags over time taking advantage of TypeScript and/or ESLint, an `ember-ultra-strict` mode, and provide tooling to expose that.
+
+A consequence of adopting `strict: true` as a behavior is that we will be progressively catching *more* type safety issues over time. Since these changes are not gated on TypeScript minor versions, they make cause “red squiggles” during any TypeScript upgrade. However, we still believe this is an appropriate default, for the following reasons:
```

Comment:
```suggestion
A consequence of adopting `strict: true` as a behavior is that we will be progressively catching *more* type safety issues over time. Since these changes are not gated on TypeScript major versions (that is, they may appear in *any* TypeScript release), they make cause “red squiggles” during any TypeScript upgrade. However, we still believe this is an appropriate default, for the following reasons:
```